lib/chef/knife/bootstrap.rb

```diff
@@ -638,8 +638,10 @@ def connect!
           EOM
           # FIXME: this should save the key to known_hosts but doesn't appear to be
           config[:ssh_verify_host_key] = :accept_new
-          do_connect(connection_opts(reset: true))
-        elsif ssh? && e.cause && e.cause.class == Net::SSH::AuthenticationFailed
+          conn_opts = connection_opts(reset: true)
+          opts.merge! conn_opts
+          retry
+        elsif (ssh? && e.cause && e.cause.class == Net::SSH::AuthenticationFailed) || (ssh? && e.class == Train::ClientError && e.message == "Your SSH Agent has no keys added, and you have not specified a password or a key file")
```

Comment:
This would work, but I'd prefer updating train [1] to populate the exception's `reason` field so that we don't have to worry about changes to text. This'll also keep moving train towards providing a consistent way to discover exception reasons.
 
[1] https://github.com/inspec/train/blob/master/lib/train/transports/ssh.rb#L106
to something like: 
```
        raise Train::ClientError.new(
            "Your SSH Agent has no keys added, and you have not specified a password or a key file",
            :no_ssh_password_or_key_available
        )
```

Reply:
Updated comparison with `e.reason`. Added PR https://github.com/inspec/train/pull/517 for this change.